src/js/components/FileInput/FileInput.js

```diff
@@ -221,198 +296,165 @@ const FileInput = forwardRef(
-          onDragLeave={() => setDragOver(false)}
-          onChange={(event) => {
-            event.persist();
-            const fileList = event.target.files;
-            const nextFiles = multiple ? [...files] : [];
-            for (let i = 0; i < fileList.length; i += 1) {
-              // avoid duplicates
-              const existing =
-                nextFiles.filter(
-                  (file) =>
-                    file.name === fileList[i].name &&
-                    file.size === fileList[i].size,
-                ).length > 0;
-              if (!existing) nextFiles.push(fileList[i]);
-            }
-            setFiles(nextFiles);
-            setDragOver(false);
-            if (onChange) onChange(event, { files: nextFiles });
-          }}
-        />
-      </ContentsBox>
+            ))}
+          <StyledFileInput
+            ref={inputRef}
+            type="file"
+            id={id}
+            name={name}
+            multiple={multiple}
+            disabled={disabled}
+            plain
+            rightOffset={rightOffset}
+            {...rest}
+            onDragOver={() => setDragOver(true)}
+            onDragLeave={() => setDragOver(false)}
+            onChange={(event) => {
+              event.persist();
+              const fileList = event.target.files;
+              let filesToAdd;
+              if (!files.length) {
+                if (multiple && max && fileList.length >= max) {
+                  const newFileList = [];
+                  for (let i = 0; i < max; i += 1) {
+                    newFileList.push(fileList[i]);
+                  }
+                  filesToAdd = newFileList;
+                } else {
+                  filesToAdd = fileList;
+                }
+              } else if (
+                multiple &&
+                max &&
+                fileList.length >= max - files.length
+              ) {
+                const newFileList = [];
+                for (let i = 0; i < max - files.length; i += 1) {
+                  newFileList.push(fileList[i]);
+                }
+                filesToAdd = newFileList;
+              } else {
+                filesToAdd = fileList;
+              }
+              const nextFiles = multiple ? [...files] : [];
+              for (let i = 0; i < filesToAdd.length; i += 1) {
+                // avoid duplicates
+                const existing =
+                  nextFiles.filter(
+                    (file) =>
+                      file.name === fileList[i].name &&
+                      file.size === fileList[i].size,
+                  ).length > 0;
+                if (!existing) nextFiles.push(filesToAdd[i]);
+              }
+              setFiles(nextFiles);
+              setDragOver(false);
+              if (onChange) onChange(event, { files: nextFiles });
+            }}
+          />
+        </ContentsBox>
+        {max && (
+          <Message {...theme.fileInput.message.maxFile}>
```

Comment:
Why the special theme for maxFile? Could it be made more generic for all messages or maybe error messages at least?

Reply:
Shall I also add one more node for the `error` messages also ?